Documentation/design-docs/first-class-structs.md

```diff
@@ -15,6 +15,271 @@ Primary Objectives
+
+  * Call args 
+    * If the struct has been promoted it is morphed to `GT_FIELD_LIST`
+      * Currently this is done only if it is passed on the stack, or if it is passed
+        in registers that exactly match the types of its fields.
+      * Proposed: This transformation would be made even if it is passed in non-matching
+        registers. The necessary transformations for correct code generation would be
+        made in `Lowering`.
+
+    * If it is passed in a single register, it is morphed into a `GT_LCL_FLD` node of the appropriate
+      type.
+      * This may involve making a copy, if the size cannot be safely loaded.
+      * Proposed: This would remain a `GT_OBJ` and would be appropriately transformed in `Lowering`,
+        e.g. using `GT_BITCAST`.
+
+    * If is passed in multiple registers
+      * A `GT_FIELD_LIST` is constructed that represents the load of each register using `GT_LCL_FLD`.
+      * Proposed: This would also remain `GT_OBJ` (or `GT_FIELD_LIST` if promoted) and would be transformed
+        to a `GT_FIELD_LIST` with the appropriate load, assemble or extraction code as needed.
+
+    * Otherwise, if it is passed by reference or on the stack, it is kept as `GT_OBJ` or `GT_LCL_VAR`
+      * Currently, if passed by reference, the value is either forced to the stack or copied.
+      * Proposed: This transformation would also be deferred until `Lowering`, at which time the
+        liveness information can provide `lastUse` information to allow a dead struct to be passed
+        directly by reference instead of being copied.
+
+It is proposed to add the following transformations in `Lowering`:
+* Transform struct values that are passed to or returned from calls by creating one or more of the following:
+  * `GT_FIELD_LIST` of `GT_LCL_FLD` when the struct is non-enregisterable and is passed in multiple
+    registers.
+  * `GT_BITCAST` when a promoted floating point field of a single-field struct is passed in an integer register.
+  * A sequence of nodes to assemble or extract promoted fields
+* Introduce copies as needed for non-last-use struct values that are passed by reference.
+
+Work Items
+----------
+This is a preliminary breakdown of the work into somewhat separable tasks.
+These work items are organized in priority order. Each work item should be able to
+proceed independently, though the aggregate effect of multiple work items may be greater
+than the individual work items alone.
+
+### Cache Struct type info
+[#21705 Pull struct type info out of GenTreeObj](https://github.com/dotnet/coreclr/pull/21705)
+  * Eliminate the use of `GT_IND` for structs
+  * Ensure that all struct-valued nodes have an index into the `ClassLayout` cache
+    * The cache includes entries for opaque "struct" values of a given size, as well as for
+      structs with known class handles
+
+### Defer ABI-specific transformations to `Lowering`.
```

Comment:
Any thoughts about getting in my removal of `GT_LIST` node from calls and field lists before doing any more significant work in call morphing and start getting conflicts left and right? :grin: It's nearly done (e.g #26392)

Reply:
Hmm, might be a bit too early to do that. I'll have to remove a ton of commits because only the first 10 or so are relevant. The rest are experiments that I wanted to pass through CI testing. Also, the ARM part is missing (I have it locally but it's buggy ATM).

Reply:
Ah, sorry about the confusion. #26392 is actually fine, I was thinking about a different PR that needs cleanup.